### PR TITLE
8260616: Removing remaining JNF dependencies in the java.desktop module

### PR DESCRIPTION
**Pull request by prrace**
This completes the desktop module JNF removal

* remove  -framework JavaNativeFoundation from make files

* remove  #import <JavaNativeFoundation/JavaNativeFoundation.h> from all source files. If needed add import of JNIUtilities.h to get jni.h definitions - better anyway since then it gets the current JDK ones not the ones from the O/S

* replace JNFNSToJavaString with NSStringToJavaString and  JNFJavaToNSString with JavaStringToNSString

* replace JNFNormalizedNSStringForPath with NormalizedPathNSStringFromJavaString and JNFNormalizedJavaStringForPath with NormalizedPathJavaStringFromNSString

* replace JNFGet/ReleaseStringUTF16UniChars with direct calls to JNI

* Map all JNFRunLoop perform* calls to the ThreadUtilities versions (the vast majority already did this)

* Redo the ThreadUtilities calls to JNFRunLoop to directly invoke NSObject perform* methods.

* define new javaRunLoopMode in ThreadUtilities to replace the JNF one and use where needed.

* Remove the single usage of JNFPerformEnvBlock

* replace JNFJavaToNSNumber in single A11Y file with local replacement

* replace single usage of JNFNSTimeIntervalToJavaMillis in ScreenMenu.m with local replacement

* remove un-needed JNFRunLoopDidStartNotification from NSApplicationAWT.m

* misc. remaining cleanup (eg missed JNF_CHECK_AND_RETHROW_EXCEPTION)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260616](https://bugs.openjdk.java.net/browse/JDK-8260616): Removing remaining JNF dependencies in the java.desktop module

### Reviewers
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to efab1de5ffd0cfa4e27ee46e37a29975ffe47558
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to efab1de5ffd0cfa4e27ee46e37a29975ffe47558
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**) ⚠️ Review applies to 8a014fa6571ac5ba6928c0033e5d8e048a591b80

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2305/head:pull/2305`
`$ git checkout pull/2305`

### CODE REVIEW

**Comment by magicus** [comment]
I mostly have questions about what is missing from this PR. :-) (If this is supposed to remove the final remnants of JNF)

- There is a disabled warning in `make/autoconf/flags-cflags.m4`, line 173, referring to JavaNativeFoundation. It can presumably be removed. If it triggers individually instead, the warning should be disabled on a per-library basis.

- In `make/modules/java.base/Lib.gmk`, line 99 & 113, are references to JavaNativeFoundation. It seems that `libosxsecurity` needs to be cleaned from JNF as well. Also, the comments indicate that the exception for STATIC_BUILD can be removed. (You can verify this with `configure --enable-static-build`)

- In `make/modules/java.desktop/Lib.gmk`, line 129, and `make/modules/java.desktop/lib/Awt2dLibraries.gmk`, line 866 & 094, it seems like `libosx`, `libawt_lwawt`, and `liboxui` also has JNF that needs to be removed. If these are fixed in any of the other issues for the umbrella JDK-8257852, I apologize. I could not figure that out.

- There is also a test dependency that I have seen being addressed, in `make/test/JtregNativeJdk.gmk` line 82, for `libTestMainKeyWindow`.

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m [hunk 1]
```diff
@@ -597,20 +598,23 @@ render using the right font (see bug 7183516) when attribString is not
...
-        JNF_CHECK_AND_RETHROW_EXCEPTION(env);
+        CHECK_EXCEPTION();
 
         // Draw the text context
         DrawTextContext(env, qsdo, awtStrike, unichars, len, x, y);
     }
     else
     {
         // Get string to draw and the length
-        const jchar *unichars = JNFGetStringUTF16UniChars(env, str);
```
Comment (gerard-ziemski) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m:608]: According to ` JNFString.h`, the `JNFGetStringUTF16UniChars()` API:

/*
 * Gets UTF16 unichars from a Java string, and checks for errors and raises a JNFException if
 * the unichars cannot be obtained from Java.
 */

raises a (JNFException) if it can't get the chars, but now we simply return if `GetStringChars()` can not return the chars.

Seems like we handle this case differently in the new code now - is this change desired and correct?

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m [hunk 2]
```diff
@@ -597,20 +598,23 @@ render using the right font (see bug 7183516) when attribString is not
     {
         jchar unichars[len];
         (*env)->GetStringRegion(env, str, 0, len, unichars);
-        JNF_CHECK_AND_RETHROW_EXCEPTION(env);
+        CHECK_EXCEPTION();
```
Comment (gerard-ziemski) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m:601]: Are 'JNF_CHECK_AND_RETHROW_EXCEPTION' and 'CHECK_EXCEPTION' equivalent?

'JNF_CHECK_AND_RETHROW_EXCEPTION' seems to throw exception, but 'CHECK_EXCEPTION' does not?

**Comment by prrace** [comment]
Right, this is the desktop module as per the first line of the comment.
java.base needs to be removed by the other PR as Erik said.
I  had not spotted it, but I don't see  why the make/test /JtregNativeJdk.gmk case needs to link this framework.
I don't see it being used by the test in question.
But we can just remove it and prove it - but probably a separate PR since it is nothing to do with the desktop module and the autoconf code needs to be updated once everything else is in.

File: src/java.desktop/macosx/native/libosxui/ScreenMenu.m [hunk 3]
```diff
@@ -156,6 +155,16 @@ - (void)handleJavaMenuItemTargetedAtIndex:(NSUInteger)menuIndex rect:(NSRect)rec
...
 }
 
+/*
+ * The input is an NSTimeInterval (a double representing seconds and fractions of seconds)
+ * 0.0 means midnight Jan 1, 2001.
+ * The output is a Java long representing time in milliseconds since midnight Jan 1st 1970.
+ * There is a Cocoa constant representing that difference : NSTimeIntervalSince1970
+ */
+static jlong NSTimeIntervalToJavaMilliseconds(NSTimeInterval interval) {
+    NSTimeInterval interval1970 = interval + NSTimeIntervalSince1970;
```
Comment (gerard-ziemski) [inline src/java.desktop/macosx/native/libosxui/ScreenMenu.m:165]: Is it required for the APIs using the value from this function to expect the time calculated since Jan 1st 1970?

File: src/java.desktop/macosx/native/libosxapp/JNIUtilities.m [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    NSString *nsStr = JavaStringToNSString(env, pathStr);
+    if (nsStr == NULL) {
+        return nil;
+    }
+    const char* chs = [nsStr fileSystemRepresentation];
+    int len = strlen(chs);
+    NSString* result = [[NSFileManager defaultManager]
+                  stringWithFileSystemRepresentation:chs length:len];
+    return result;
+}
```
Comment (gerard-ziemski) [inline src/java.desktop/macosx/native/libosxapp/JNIUtilities.m:83]: Why are we doing:

`java_string -> chars -> NSString -> chars -> [NSFileManager stringWithFileSystemRepresentation]`

?

Why not just get the raw characters form JNI and feed them directly into `[NSFileManager  stringWithFileSystemRepresentation]`, ie:

`java_string -> chars -> [NSFileManager stringWithFileSystemRepresentation]`

and skip the `JavaStringToNSString` step altogether?

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]

**Comment by gerard-ziemski** [comment]
Lots of small changes you had to handle here. Thank you for doing it!

**Comment by prrace** [comment]
I pushed a commit with the remaining -lframework ... removals in the desktop makefiles that I had somehow missed ...

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m (hunk 2 above)
Comment (prrace) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m:601]: JNF_CHECK_AND_RETHROW_EXCEPTION is this

/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/System/Library/Frameworks/JavaNativeFoundation.framework/Versions/A/Headers/JNFJNI.h

// JNF_CHECK_AND_RETHROW_EXCEPTION - rethrows exceptions from Java
//
// Takes an exception thrown from Java, and transforms it into an
// NSException. The NSException should bubble up to the upper-most
// JNF_COCOA_ENTER/JNF_COCOA_EXIT pair, and then be re-thrown as
// a Java exception when returning from JNI. This check should be
// done after raw JNI operations which could cause a Java exception
// to be be thrown. The JNF{Get/Set/Call}  macros below do this
// check automatically.
#define JNF_CHECK_AND_RETHROW_EXCEPTION(env)                                                    \
{                                                                                                                                       \
    jthrowable _exception = (*env)->ExceptionOccurred(env);                     \
    if (_exception) [JNFException raise:env throwable:_exception];      \
}

So what it actually does is clear the exception, raise an NSException and when
the code reaches JNF_COCOA_EXIT - which then has to be there - it clears the NSException
and re-throws the Java exception.

So the name of the macro is misleading since this does NOT itself rethrow the exception,
it relies on other code to do that.

CHECK_EXCEPTION will amount to the same - it throws an NSException if there is a Java exception pending.
And it will clear the NSException before exiting back to Java.
The difference is that it doesn't clear the Java Exception and later rethrow it since there is no need
There is one exception to this in  CHECK_EXCEPTION - if this is the main (ie appkit) thread the java exception is cleared since there's no one to return it to ...

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m (hunk 1 above)
Comment (prrace) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m:608]: You are correct, but I decided that here it is fine.
If GetStringChars fails it will return NULL (it does not raise any excptions itself).
I added a check for NULL if NULL we then return from the JNI method to Java.
In the old case it also would return but would additionally propagate that raised exception to Java which JNF
decided to throw into the mix when there's already a standard way of testing failure.
And since we already know str != NULL we are in the realms of something went badly wrong inside the VM for
this to fail. So I decided this was all fine.

File: src/java.desktop/macosx/native/libosxapp/JNIUtilities.m (hunk 4 above)
Comment (prrace) [reply src/java.desktop/macosx/native/libosxapp/JNIUtilities.m:83]: I tried to explain the odd-ness of this in the comments preceding the function definition.
Near as I could figure out that intermediate call is needed to do the decomposition since the NSFileManager won't do that.
But this is not exactly my area of expertise and there may be a better way. Who is an expert on the intricacies of the macOS file system naming ?

**Review by magicus — APPROVED** [review]
Build changes look good for this PR.

**Comment by prrace** [comment]
> * There is also a test dependency that I have seen being addressed, in `make/test/JtregNativeJdk.gmk` line 82, for `libTestMainKeyWindow`.

update on this. It Is used by precisely one jtreg AWT test. So I guess it is fair game to add resolving this to the desktop module update. I am currently testing it before committing - will take a couple of hours as the headful tests take a while to run.

File: src/java.desktop/macosx/native/libosxapp/JNIUtilities.m (hunk 4 above)
Comment (gerard-ziemski) [reply src/java.desktop/macosx/native/libosxapp/JNIUtilities.m:83]: How about a comment like:

`/*
 Returns an NSString in decomposed UTF16 format that is compatable with HFS's
 expectation of the UTF16 format for file system paths.
 
 Example string: "/Users/Amélie/"
 
 Java's UTF16 string is                 "/ U s e r s / A m \351 l i e /"
 macOS UTF16 string suitable for HFS is "/ U s e r s / A m e \314 \201 l i e /"
 
 There is no direct API that takes in NSString UTF16 encoded by Java
 and produces NSString UTF16 for HFS, so we first need to decompose it
 into chars (suitable for low level C file APIs), and only then
 create NSString representation of this decomposition back into UTF16 string.
 */`

?

I borrowed the API description from Apple's original `JNFNormalizedNSStringForPath` API, but added an example and different description.

**Review by gerard-ziemski — APPROVED** [review]

**Comment by gerard-ziemski** [comment]
The changes look good to me, but please see my comment from my review about documenting `NormalizedPathNSStringFromJavaString`

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m [hunk 5]
```diff
@@ -597,20 +598,23 @@ render using the right font (see bug 7183516) when attribString is not
...
         // Draw the text context
         DrawTextContext(env, qsdo, awtStrike, unichars, len, x, y);
     }
     else
     {
         // Get string to draw and the length
-        const jchar *unichars = JNFGetStringUTF16UniChars(env, str);
+        const jchar *unichars = (*env)->GetStringChars(env, str, NULL);
+        if (unichars == NULL) {
+            return;
```
Comment (mrserb) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m:611]: Do not we need to throw an exception here? Otherwise, GetStringChars error will be ignored?

File: src/java.desktop/macosx/native/libawt_lwawt/awt/JavaComponentAccessibility.m [hunk 6]
```diff
@@ -956,6 +955,32 @@ - (BOOL)accessibilityIsIndexAttributeSettable
...
+/*
+ * The java/lang/Number concrete class could be for any of the Java primitive
+ * numerical types or some other subclass.
+ * All existing A11Y code uses Integer so that is what we look for first
+ * But all must be able to return a double and NSNumber accepts a double,
+ * so that's the fall back.
+ */
+static NSNumber* JavaNumberToNSNumber(JNIEnv *env, jobject jnumber) {
+    if (jnumber == NULL) {
+        return nil;
```
Comment (mrserb) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/JavaComponentAccessibility.m:967]: Based on its usage it is probably should be zero on NULL number?

File: src/java.desktop/macosx/native/libosxapp/JNIUtilities.m [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "JNIUtilities.h"
+
+NSString* JavaStringToNSString(JNIEnv *env, jstring jstr) {
+    if (jstr == NULL) {
+        return NULL;
```
Comment (mrserb) [inline src/java.desktop/macosx/native/libosxapp/JNIUtilities.m:30]: In other methods, the nil is used

File: src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m [hunk 8]
```diff
@@ -49,6 +50,15 @@ static inline void attachCurrentThread(void** env) {
 
 @implementation ThreadUtilities
 
++ (void)initialize {
```
Comment (mrserb) [inline src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m:53]: I think we need to check how this new modes will work when the AWT is embedded inside SWT and FX.

File: src/java.desktop/macosx/native/libawt_lwawt/awt/JavaComponentAccessibility.m (hunk 6 above)
Comment (prrace) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/JavaComponentAccessibility.m:967]: Not an unreasonable idea and I considered it but :
- It is never called with NULL. There is always a null check
- The JNF equivalent returns nil on NULL

BTW two of the functions in which the code appears : accessibilityMinValueAttribute and accessibilityMaxValueAttribute  (SFAIC) aren't used anywhere.

File: src/java.desktop/macosx/native/libosxapp/JNIUtilities.m (hunk 7 above)
Comment (prrace) [reply src/java.desktop/macosx/native/libosxapp/JNIUtilities.m:30]: Not understanding what you mean ? This is the same behavior as the function being replaced.

File: src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m (hunk 8 above)
Comment (prrace) [reply src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m:53]: We are just specifying an additional run mode for JDK internal use.
It means that when we are saying to process only events for that mode, then only those will be processed.
And it is used only for nested event loops.
Nothing eternal should be assuming AWT uses JNF at all, never mind in a particular way.

There is a special case for FX but I don't see a problem.

If you look at Java_sun_lwawt_macosx_LWCToolkit_doAWTRunLoopImpl
there's a variable "inAWT".
         isRunning = [[NSRunLoop currentRunLoop] runMode:(inAWT ? [JNFRunLoop javaRunLoopMode] : NSDefaultRunLoopMode) ... ]];

This is specified as
        inAWT = AccessController.doPrivileged(new PrivilegedAction<Boolean>() {
            @Override
            public Boolean run() {
                return !Boolean.parseBoolean(System.getProperty("javafx.embed.singleThread", "false"));
            }
        });
    }

So generally FX doesn't care, and is ignorant of this new mode.
Unless you set that property to true, in which case AWT use the  NSDefaultRunLoopMode and so again FX is unaffected.
Nothing in the FX sources goes anywhere near JNF .. or has a reference to the special mode.

Bottom line I don't see it being affected. 

I'll check with Kevin but also Gerard had a lot to do with the creation of the current FX toolkit.

Comment (kevinrushforth) [reply src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m:53]: I ran some tests embedding JavaFX into Swing and vice versa both with and without `-Djavafx.embed.singleThread=true` and I don't see any regression in behavior.

**Comment by kevinrushforth** [comment]
The following commit was integrated to jdk master since you created this branch:

acbcde8 : [JDK-8256111](https://bugs.openjdk.java.net/browse/JDK-8256111) : Create implementation for NSAccessibilityStaticText protocol

with a new reference to JNF, so it fails to compile with this patch. You will need to merge openjdk/jdk master into your PR branch and then fix the new JNF reference.

File: src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m (hunk 8 above)
Comment (mrserb) [reply src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m:53]: I am mostly worried about the usage of JNF by someone else's native code, as far as I understand it could be "broken" now. But it is good that FX does not use it.

BTW looks like all comments like  "// AWT_THREADING Safe (AWTRunLoop)" could be removed now.

Comment (prrace) [reply src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m:53]: SWT does not use JNF at all. I've looked at their source code.
I can clean up those comments.

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m (hunk 5 above)
Comment (prrace) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m:611]: > The changes look good to me, but please see my comment from my review about documenting `NormalizedPathNSStringFromJavaString()` API.

I see it. I will copy what you wrote in there.

Comment (prrace) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/CTextPipe.m:611]: > I read it and not sure that it is fine to ignore this error, why not throw an exception and signal the CTextPipe_doDrawString that an error occurred like InvalidPipeException or something(Sometimes we wrap other exception like OOM into the InvalidPipeException and this seems similar case)?

OK. I will do something like throw OOME or InternalError. Since we already checked for NULL as an arg any failure here implies a deep VM problem rather than anytjhing like a 2D invalid pipe

**Comment by prrace** [comment]
Updated as follows
- merged to current master and fixed the new A11Y usage of JNF
- added comment provided by Gerard
- removed comments as suggested by Sergey
- made failure of GetStringChars in CTextPipe throw OOME

**Review by mrserb — APPROVED** [review]
